spec/facter/resolvers/linux/hostname_spec.rb

```diff
@@ -0,0 +1,254 @@
+# frozen_string_literal: true
+
+describe Facter::Resolvers::Linux::Hostname do
+  subject(:hostname_resolver) { Facter::Resolvers::Linux::Hostname }
+
+  let(:log_spy) { instance_spy(Facter::Log) }
+
+  describe '#resolve' do
+    before do
+      hostname_resolver.instance_variable_set(:@log, log_spy)
+      allow(Socket).to receive(:gethostname).and_return(host)
+      allow(Facter::Util::FileHelper).to receive(:safe_read)
+        .with('/etc/resolv.conf')
+        .and_return("nameserver 10.10.0.10\nnameserver 10.10.1.10\nsearch baz\ndomain baz\n")
+    end
+
+    after do
+      hostname_resolver.invalidate_cache
+    end
+
+    context 'when ruby socket hostname works' do
+      let(:hostname) { 'foo' }
+      let(:domain) { 'bar' }
+      let(:fqdn) { "#{hostname}.#{domain}" }
+
+      context 'when it returns fqdn' do
+        let(:host) { "#{hostname}.#{domain}" }
+
+        it 'detects hostname' do
```

Comment:
You are performing the exact same checks in every scenario. I think that adding these three checks inside a shared_example would greatly improve readability of the different scenarios.